src/mlpack/methods/ann/layer/nearest_interpolation_impl.hpp

```diff
@@ -0,0 +1,172 @@
+{
+  // Nothing to do here. 
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>::
+NearestInterpolationType(NearestInterpolationType&& other) :
+  Layer<MatType>(std::move(other)),
+  scaleFactors(std::move(other.scaleFactors))
+{
+  // Nothing to do here.
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>&
+NearestInterpolationType<MatType>::
+operator=(const NearestInterpolationType& other) 
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(other);
+    scaleFactors = other.scaleFactors;
+  }
+  return *this;
+}
+
+template<typename MatType>
+NearestInterpolationType<MatType>&
+NearestInterpolationType<MatType>::
+operator=(NearestInterpolationType&& other) 
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    scaleFactors = std::move(other.scaleFactors);
+  }
+  return *this;
+}
+
+template<typename MatType>
+void NearestInterpolationType<MatType>::Forward(
+  const MatType& input, MatType& output)
+{
+  const size_t channels = this->inputDimensions[2];
+
+  const size_t outRowSize = this->outputDimensions[0];
+  const size_t outColSize = this->outputDimensions[1];
+
+  const size_t inRowSize = this->inputDimensions[0];
+  const size_t inColSize = this->inputDimensions[1];
+
+  arma::cube inputAsCube;
+  arma::cube outputAsCube;
+
+  MakeAlias(inputAsCube, input, inRowSize, inColSize, channels, 0, false);
+  MakeAlias(outputAsCube, output, outRowSize, outColSize, channels, 0, true);
+
+  for (size_t i = 0; i < outRowSize; ++i)
+  {
+    size_t rOrigin = std::floor(i  / scaleFactors[0]);
+    for (size_t j = 0; j < outColSize; ++j)
+    {
+      size_t cOrigin = std::floor(j / scaleFactors[1]);
+      for (size_t k = 0; k < channels; ++k)
+      {
+        outputAsCube(i, j, k) = inputAsCube(rOrigin, cOrigin, k);
+      }
+    }
+  }
+}
+
+template<typename MatType>
+void NearestInterpolationType<MatType>::Backward(
+  const MatType& /*input*/,
+  const MatType& gradient,
+  MatType& output)
+{
+  const size_t channels = this->inputDimensions[2];
+
+  const size_t outRowSize = this->outputDimensions[0];
+  const size_t outColSize = this->outputDimensions[1];
+
+  const size_t inRowSize = this->inputDimensions[0];
+  const size_t inColSize = this->inputDimensions[1];
+
+  arma::cube outputAsCube;
+  arma::cube gradientAsCube;
+
+  MakeAlias(outputAsCube, output, inRowSize, inColSize, channels, 0, true);
+  MakeAlias(gradientAsCube, gradient, outRowSize, outColSize, channels, 0, false);
+
+  for (size_t i = 0; i < outRowSize; ++i)
+  {
+    size_t rOrigin = std::floor(i / scaleFactors[0]);
+    for (size_t j = 0; j < outColSize; ++j)
+    {
+      size_t cOrigin = std::floor(j / scaleFactors[1]);
+      for (size_t k = 0; k < channels; ++k)
+      {
+        outputAsCube(rOrigin, cOrigin, k) += gradientAsCube(i, j, k);
+      }
+    }
+  }
+}
+
+template<typename MatType>
+void NearestInterpolationType<MatType>::ComputeOutputDimensions()
```

Comment:
This makes some number of assumptions that I don't think we need to make.  Specifically, it requires that the input dimension size is exactly 3, but this is not what the code requires.  We only need that the number of dimensions is at least equal to 2, then we apply the scaling to the first two dimensions and leave the rest unchanged (that is what the input code does).